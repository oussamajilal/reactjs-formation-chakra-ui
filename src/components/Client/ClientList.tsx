import * as React from "react";
import { Button } from "@chakra-ui/button";
import { PlusSquareIcon } from "@chakra-ui/icons";
import { Container, Heading } from "@chakra-ui/layout";
import ClientTable from "./ClientTable";
import { useDisclosure } from "@chakra-ui/hooks";
import NewClientModal from "./NewClientModal";
import { Client } from "./Client";
import { useQuery } from "react-query";
import queries from "../queries";
import Loader from "../Loader";
import ErrorAlert from "../ErrorAlert";
import { UserContext } from "../UserContext";

const ClientList = () => {
  const { user } = React.useContext(UserContext);
  const { data, error, isError, isLoading, refetch } = useQuery(
    "clients",
    queries.fetchClients
  );

  const { isOpen, onOpen, onClose } = useDisclosure();

  if (isLoading) {
    return <Loader />;
  }

  if (isError && error instanceof Error) {
    return <ErrorAlert message={error.message} />;
  }

  const handleSubmit = () => {
    refetch();
  };

  return (
    <>
      {user && (
        <Heading>
          Welcome {user.firstName} {user.lastName}
        </Heading>
      )}
      {data && (
        <>
          <ClientTable clients={data} />
          <Container maxW="container.lg" paddingTop="5">
            <Button
              data-testid="modal-btn"
              leftIcon={<PlusSquareIcon />}
              colorScheme="green"
              variant="solid"
              onClick={onOpen}
            >
              Add a new client
            </Button>
            <NewClientModal
              isOpen={isOpen}
              onClose={onClose}
              handleSubmit={handleSubmit}
            />
          </Container>
        </>
      )}
    </>
  );
};

export default ClientList;
