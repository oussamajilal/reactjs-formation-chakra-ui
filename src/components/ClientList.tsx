import * as React from "react";
import { Button } from "@chakra-ui/button";
import { PlusSquareIcon } from "@chakra-ui/icons";
import { Container } from "@chakra-ui/layout";
import ClientTable from "./ClientTable";
import { useDisclosure } from "@chakra-ui/hooks";
import NewClientModal from "./NewClientModal";
import { Client } from "./Client";

const ClientList = () => {
  const initialClients: Array<Client> = [
    {
      id: "1",
      firstName: "Ahmed",
      lastName: "Ali",
      email: "ahmed.ali@email.com",
      phoneNumber: "+212666778899",
      height: 72.5,
      isActive: true,
    },
    {
      id: "2",
      firstName: "Salim",
      lastName: "Karam",
      email: "salim.karam@email.com",
      phoneNumber: "+212666778898",
      height: 65.0,
      isActive: false,
    },
    {
      id: "3",
      firstName: "Soukaina",
      lastName: "Latif",
      email: "soukaina.latif@email.com",
      phoneNumber: "+212666778897",
      height: 69.3,
      isActive: true,
    },
  ];

  const [clients, setClients] = React.useState(initialClients);

  const { isOpen, onOpen, onClose } = useDisclosure();

  const handleSubmit = (newClient: Client) => {
    setClients([...clients, newClient]);
  };

  return (
    <>
      <ClientTable clients={clients} />
      <Container maxW="container.lg" paddingTop="5">
        <Button
          data-testid="modal-btn"
          leftIcon={<PlusSquareIcon />}
          colorScheme="green"
          variant="solid"
          onClick={onOpen}
        >
          Add a new client
        </Button>
        <NewClientModal
          isOpen={isOpen}
          onClose={onClose}
          handleSubmit={handleSubmit}
        />
      </Container>
    </>
  );
};

export default ClientList;
