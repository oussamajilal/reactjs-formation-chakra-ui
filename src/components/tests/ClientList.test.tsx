import { fireEvent } from "@testing-library/dom";
import * as React from "react";
import { render } from "../../test-utils";
import ClientList from "../ClientList";

describe("ClientList", () => {
  it("should render correctly", () => {
    const componentDom = render(<ClientList />);
    expect(componentDom.asFragment()).toMatchSnapshot();
  });

  it("should render modal when modal-btn is clicked", () => {
    const componentDom = render(<ClientList />);
    const modalBtn = componentDom.getByTestId("modal-btn");
    modalBtn.click();
    const form = componentDom.getByTestId("new-client-form");
    expect(form).toBeInTheDocument();
    const firstNameInput = componentDom.getByTestId("new-client-first-name");
    fireEvent.change(firstNameInput, {target: {value: 'fake first name'}})
    const lastNameInput = componentDom.getByTestId("new-client-last-name");
    fireEvent.change(lastNameInput, {target: {value: 'fake last name'}})

    // TODO: fill the entire form
    
    const formSubmit = componentDom.getByTestId("newclient-form-submit");
    formSubmit.click();
    //expect()
  });
});
