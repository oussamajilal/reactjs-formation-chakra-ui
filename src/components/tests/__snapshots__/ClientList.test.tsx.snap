// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientList should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-container css-1xnqsi7"
  >
    <table
      class="chakra-table css-5605sr"
      role="table"
    >
      <caption
        class="css-7478fk"
      >
        Users List
      </caption>
      <thead
        class="css-0"
      >
        <tr
          class="css-0"
          role="row"
        >
          <th
            class="css-1sa2h46"
          >
            ID
          </th>
          <th
            class="css-1sa2h46"
          >
            First Name
          </th>
          <th
            class="css-1sa2h46"
          >
            Last Name
          </th>
          <th
            class="css-1sa2h46"
          >
            Email
          </th>
          <th
            class="css-1sa2h46"
          >
            Phone Number
          </th>
          <th
            class="css-1sa2h46"
            data-is-numeric="true"
          >
            Height
          </th>
          <th
            class="css-1sa2h46"
          >
            Active
          </th>
        </tr>
      </thead>
      <tbody
        class="css-woqubx"
      >
        <tr
          class="css-0"
          role="row"
        >
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            1
          </td>
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            Ahmed
          </td>
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            Ali
          </td>
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            ahmed.ali@email.com
          </td>
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            +212666778899
          </td>
          <td
            class="css-12y4u6o"
            data-is-numeric="true"
            role="gridcell"
          >
            72.5
          </td>
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            <svg
              class="chakra-icon css-y3r3qv"
              focusable="false"
              viewBox="0 0 14 14"
            >
              <g
                fill="currentColor"
              >
                <polygon
                  points="5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
                />
              </g>
            </svg>
          </td>
        </tr>
        <tr
          class="css-0"
          role="row"
        >
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            2
          </td>
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            Salim
          </td>
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            Karam
          </td>
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            salim.karam@email.com
          </td>
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            +212666778898
          </td>
          <td
            class="css-12y4u6o"
            data-is-numeric="true"
            role="gridcell"
          >
            65
          </td>
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            <svg
              class="chakra-icon css-dzn7nc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8 0-1.85.63-3.55 1.69-4.9L16.9 18.31C15.55 19.37 13.85 20 12 20zm6.31-3.1L7.1 5.69C8.45 4.63 10.15 4 12 4c4.42 0 8 3.58 8 8 0 1.85-.63 3.55-1.69 4.9z"
                fill="currentColor"
              />
            </svg>
          </td>
        </tr>
        <tr
          class="css-0"
          role="row"
        >
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            3
          </td>
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            Soukaina
          </td>
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            Latif
          </td>
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            soukaina.latif@email.com
          </td>
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            +212666778897
          </td>
          <td
            class="css-12y4u6o"
            data-is-numeric="true"
            role="gridcell"
          >
            69.3
          </td>
          <td
            class="css-12y4u6o"
            role="gridcell"
          >
            <svg
              class="chakra-icon css-y3r3qv"
              focusable="false"
              viewBox="0 0 14 14"
            >
              <g
                fill="currentColor"
              >
                <polygon
                  points="5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
                />
              </g>
            </svg>
          </td>
        </tr>
      </tbody>
      <tfoot
        class="css-5wov5z"
      >
        <tr
          class="css-0"
          role="row"
        >
          <th
            class="css-1sa2h46"
          >
            ID
          </th>
          <th
            class="css-1sa2h46"
          >
            First Name
          </th>
          <th
            class="css-1sa2h46"
          >
            Last Name
          </th>
          <th
            class="css-1sa2h46"
          >
            Email
          </th>
          <th
            class="css-1sa2h46"
          >
            Phone Number
          </th>
          <th
            class="css-1sa2h46"
            data-is-numeric="true"
          >
            Height
          </th>
          <th
            class="css-1sa2h46"
          >
            Active
          </th>
        </tr>
      </tfoot>
    </table>
  </div>
  <div
    class="chakra-container css-1m2doq5"
  >
    <button
      class="chakra-button css-5w5n0l"
      data-testid="modal-btn"
      type="button"
    >
      <span
        class="chakra-button__icon css-1wh2kri"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <g
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-width="2"
          >
            <rect
              height="18"
              rx="2"
              ry="2"
              width="18"
              x="3"
              y="3"
            />
            <path
              d="M12 8v8"
            />
            <path
              d="M8 12h8"
            />
          </g>
        </svg>
      </span>
      Add a new client
    </button>
  </div>
</DocumentFragment>
`;
